source/state_representation/test/tests/test_parameter.cpp

```diff
@@ -84,3 +87,66 @@ TEST(ParameterTest, MakeShared) {
   EXPECT_EQ(param->get_value().get_reference_frame(), "B");
   EXPECT_TRUE(param->get_value().data().isApprox(pose.data()));
 }
+
+TEST(ParameterTest, ParameterThroughInterface) {
+  auto pose = CartesianPose::Random("A", "B");
+  std::shared_ptr<ParameterInterface> param_interface = make_shared_parameter("name", pose);
+
+  auto param = param_interface->get_parameter<CartesianPose>();
+  EXPECT_EQ(param->get_name(), "name");
+  EXPECT_EQ(param->get_type(), StateType::PARAMETER_CARTESIANPOSE);
+  EXPECT_EQ(param->is_empty(), false);
+  EXPECT_EQ(param->get_value().get_name(), "A");
+  EXPECT_EQ(param->get_value().get_reference_frame(), "B");
+  EXPECT_TRUE(param->get_value().data().isApprox(pose.data()));
+
+  auto param_value = param_interface->get_parameter_value<CartesianPose>();
+  EXPECT_EQ(param_value.get_name(), "A");
+  EXPECT_EQ(param_value.get_reference_frame(), "B");
+  EXPECT_TRUE(param_value.data().isApprox(pose.data()));
+
+  auto pose2 = CartesianPose::Random("C", "D");
+  param_interface->set_parameter_value(pose2);
+  param_value = param_interface->get_parameter_value<CartesianPose>();
+  EXPECT_EQ(param_value.get_name(), "C");
+  EXPECT_EQ(param_value.get_reference_frame(), "D");
+  EXPECT_TRUE(param_value.data().isApprox(pose2.data()));
+}
+
+TEST(ParameterTest, ParameterInterfaceBadPointer) {
+  ParameterInterface parameter_interface(
+      StateType::PARAMETER_INT,
+  "name");
+  EXPECT_THROW(parameter_interface.get_parameter<int>(), std::exception); // bad pointer
+  EXPECT_THROW(parameter_interface.get_parameter<int>(true), std::exception); // bad pointer
+  EXPECT_THROW(parameter_interface.get_parameter<int>(false), std::exception); // bad pointer
```

Comment:
why does this still throw?

Reply:
My intention for the `validate_pointer` flag was to handle the null pointer return with an exception _if_ the dynamic cast could not resolve. However, this fails earlier because this `ParameterInterface` object is being directly accessed and is not owned by any shared pointers (its reference count is zero), meaning we can't use `dynamic_pointer_cast` and also shouldn't used `shared_from_this`. To be fair I could add a `try` to this first line and return null pointer if it fails and the flag is `false`.